Process abort in state tests on some macOS versions with Rust 1.70

## Tasks

- [x] #7834 
- [x] #7842
  - [x] Admin: Add a macOS test branch protection rule

## Motivation

Zebra's state tests are crashing on macOS across multiple PRs:

<details>

> Running `/Users/runner/work/zebra/zebra/target/release/deps/zebra_state-52d6a4509c1a5e9a`
> running 97 tests
> test service::chain_tip::tests::vectors::chain_tip_change_is_initially_not_ready ... ok
> test service::chain_tip::tests::vectors::current_best_tip_is_initially_empty ... ok
> test service::chain_tip::tests::vectors::empty_latest_chain_tip_is_empty ... ok
> test service::chain_tip::tests::prop::best_tip_is_latest_non_finalized_then_latest_finalized ... ok
> 2023-06-01T20:21:10.039357Z  INFO Opened Zebra state cache at /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/zebra-state-v25-mainnet-T3MWk1
> 2023-06-01T20:21:10.039660Z  INFO loaded Zebra state cache tip=None
> 2023-06-01T20:21:10.040026Z  INFO Opened Zebra state cache at /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/zebra-state-v25-mainnet-n6VK4c
> 2023-06-01T20:21:10.040049Z  INFO loaded Zebra state cache tip=None
> 2023-06-01T20:21:10.130630Z  INFO Opened Zebra state cache at /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/zebra-state-v25-mainnet-OSJIVt
> 2023-06-01T20:21:10.130704Z  INFO loaded Zebra state cache tip=None
> error: test failed, to rerun pass `-p zebra-state --lib`
> Caused by:

</details>

>   process didn't exit successfully: `/Users/runner/work/zebra/zebra/target/release/deps/zebra_state-52d6a4509c1a5e9a` (signal: 6, SIGABRT: process abort signal)
> Error: Process completed with exit code 101.

https://github.com/ZcashFoundation/zebra/actions/runs/5148228655/jobs/9269653348#step:15:3285
https://github.com/ZcashFoundation/zebra/actions/runs/5151593461/jobs/9276906631?pr=6810#step:15:3274

### Investigation

This appears to be an environment issue, because it wasn't failing on the `main` branch when that code was originally merged:
https://github.com/ZcashFoundation/zebra/actions/runs/5144319422/jobs/9260464558

Runners started being updated to new software versions on June 2:
https://github.com/actions/runner-images/issues/7660

The macOS and image versions on failing and successful runners are the same:
> macOS 12.6.5 21G531

https://github.com/ZcashFoundation/zebra/actions/runs/5144319422/jobs/9260464558#step:1:4
https://github.com/ZcashFoundation/zebra/actions/runs/5151593461/jobs/9276906631?pr=6810#step:1:4

But the Rust versions are different:
> stable-x86_64-apple-darwin unchanged - rustc 1.69.0 (84c898d65 2023-04-16)

https://github.com/ZcashFoundation/zebra/actions/runs/5144319422/jobs/9260464558#step:5:18

> stable-x86_64-apple-darwin updated - rustc 1.70.0 (90c541806 2023-05-31) (from rustc 1.69.0 (84c898d65 2023-04-16))

https://github.com/ZcashFoundation/zebra/actions/runs/5151593461/jobs/9276906631?pr=6810#step:5:34

Teor can't reproduce this bug on their local machines with Rust 1.70, so it might be processor or OS-version dependent:
- macOS Ventura 13.2 on M1 running aarch64 code
- macOS Ventura 13.2 on M1 emulating x86_64 code 
- Linux on x86_64

### Diagnosis

We might need to make Zebra compatible with Rust 1.70 and later on macOS 12 and earlier.

We could disable these tests until we do, because macOS is not a supported Zebra platform.

### Complex Code or Requirements

Usually this happens in the state due to RocksDB or state service shutdown or concurrency bugs.

### Testing

Our existing tests seem to reliably detect this bug.

--- a/.github/workflows/ci-unit-tests-os.patch.yml
+++ b/.github/workflows/ci-unit-tests-os.patch.yml
@@ -21,7 +21,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        os: [ubuntu-latest]
+        os: [ubuntu-latest, macos-latest]
         rust: [stable, beta]
         features: [""]
         exclude:

--- a/.github/workflows/ci-unit-tests-os.yml
+++ b/.github/workflows/ci-unit-tests-os.yml
@@ -74,7 +74,6 @@ jobs:
       fail-fast: false
       matrix:
         # TODO: Windows was removed for now, see https://github.com/ZcashFoundation/zebra/issues/3801
-        # TODO: macOS tests were removed for now, see https://github.com/ZcashFoundation/zebra/issues/6824
         os: [ubuntu-latest, macos-latest]
         rust: [stable, beta]
         # TODO: When vars.EXPERIMENTAL_FEATURES has features in it, add it here.

--- a/.github/workflows/ci-unit-tests-os.yml
+++ b/.github/workflows/ci-unit-tests-os.yml
@@ -75,7 +75,7 @@ jobs:
       matrix:
         # TODO: Windows was removed for now, see https://github.com/ZcashFoundation/zebra/issues/3801
         # TODO: macOS tests were removed for now, see https://github.com/ZcashFoundation/zebra/issues/6824
-        os: [ubuntu-latest]
+        os: [ubuntu-latest, macos-latest]
         rust: [stable, beta]
         # TODO: When vars.EXPERIMENTAL_FEATURES has features in it, add it here.
         #       Or work out a way to trim the space from the variable: GitHub doesn't allow empty variables.

